site/app/repositories/email/EmailRepository.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+namespace app\repositories\email;
+
+use app\entities\email\EmailEntity;
+use app\models\EmailStatusModel;
+use Doctrine\ORM\EntityRepository;
+use Doctrine\ORM\Tools\Pagination\Paginator;
+
+class EmailRepository extends EntityRepository {
+    const PAGE_SIZE = 1000;
+    const MAX_SUBJECTS_PER_PAGE = 10;
+
+    public function getEmailsByPage(int $page, $semester = null, $course = null): Paginator {
+        $entity = EmailEntity::class;
+        $course_specific = ($semester && $course) ? "WHERE e.semester = '{$semester}' AND e.course = '{$course}' " : "";
+        $dql = "SELECT e FROM {$entity} e {$course_specific} ORDER BY e.created DESC";
+        $window = $this->getPageWindow($page, $semester, $course);
+        $query = $this->_em->createQuery($dql)
+            ->setFirstResult($window[0])
+            ->setMaxResults($window[1]);
+            //->getResult(); maybe
+
+        return new Paginator($query);
+    }
+
+    public function getPageNum($semester = null, $course = null): int {
+        $entity = EmailEntity::class;
+        if ($semester != null || $course != null) {
+            $dql = "SELECT e.subject, e.created, COUNT(e) FROM {$entity} e WHERE e.semester = '{$semester}' AND e.course = '{$course}' GROUP BY e.subject, e.created ORDER BY e.created DESC";
+        }
+        else {
+            $dql = "SELECT e.subject, e.created, COUNT(e) FROM {$entity} e GROUP BY e.subject, e.created ORDER BY e.created DESC";
+        }
+        $q = $this->_em->createQuery($dql);
+        $query_res = $q->getResult();
+        $page = 1;
+        $count = 0;
+        $subject = 0;
+        foreach ($query_res as $email) {
+            $count += $email[1];
+            $subject += 1;
+            if ($count > self::PAGE_SIZE || $subject >= self::MAX_SUBJECTS_PER_PAGE) {
```

Comment:
What is $count is 3x the size of PAGE_SIZE shouldn't $page be increased more than 1?

Reply:
When I first thought of this, I thought that that if it increases by more than 1 then we will have to split the same email (same subject, same created timestamp) into multiple pages which might cause confusion, so I decided to make 10,000 a softer limit for making it a new page.